### PR TITLE
Improve screen jumping algorithm

### LINKED ISSUE #16: Adjacent monitor jump issue
Hi there,

This is a lovely tool and I just installed it on a new device. I found some unexpected behaviour in what I guess comes from the `+j` option for jumping the mouse cursor to adjacent monitors that aren't quite aligned with each other.

I've added a StartUp link that runs the `.exe` as follows:
```
C:\Users\<username>\.tools\MouseUnSnag\MouseUnSnag.exe +s +j -w
```

Let me illustrate the issue by showing my monitor setup and adding some arrows, like in the readme.

**Expected behaviour:**
image
The cursor moving along the top of screen 3 onto screen 4 should be moved from the top-right of screen 3 to the top-left of screen 4.
Similarly, the cursor moving along the top of screen 2 onto screen 4 should be moved from the top-left of screen 2 to the top-right of screen 4.

**Observed behaviour:**
image
The cursor jumps from the top-right of screen 3 to the top-left of screen 1.
Similarly, the cursor jumps from the top-left of screen 2 to the top-right of screen 1.

I tried to work around this by just disabling the 'Jump between monitors' option, but that will have the mouse snag in my Screen3-top-right and Screen2-top-left corners and not move to another screen at all.

### PR DESCRIPTION
**Pull request by GrantMoyer**
The current algorithm is roughly:
1. For each monitor in the same order the OS numbers them
    1. If the cursor moved off the current monitor horizontally (including if it also moved vertically, i.e. diagonally), and the monitor's edge lines up exactly with the current monitor, jump to that monitor
    2. if the cursor moved of the current monitor vertically, and the monitor's edge lines up exactly with the current monitor, jump to that monitor

This causes errors, as shown in issue #16, when more than one monitor is in the direction the cursor leaves the screen. Instead we'd like to pick the "best" screen in the direction the cursor leaves the screen. "Best" isn't well defined, but closest by euclidean distance should be a reasonable approximation.

The new algorithm is:
1. Filter the list of all monitors to just those in the direction the cursor left the current monitor.
2. Jump to the monitor in the filtered list which is closest to where the cursor left the current monitor.

Fixes #16

### CODE REVIEW

**Comment by GrantMoyer** [comment]
Since I can't test these changes locally in my linux dev environment (MouseUnSnag.Win32Interop.NativeMethods links functions from windows dlls), I prefer to wait to merge this until #15 is merged, so this can be tested with GitHub actions.

File: MouseUnSnag/GeometryUtil.cs
Comment (Robin-S) [inline MouseUnSnag/GeometryUtil.cs:105]: To account for different combinations of the `OverlapX` and `OverlapY` results, using a `switch` expression seems a bit convoluted. What about this?
```
var x = OverlapX(r1, r2) ? 0 : Math.Sign(r1.Left - r2.Left);
var y = OverlapY(r1, r2) ? 0 : Math.Sign(r1.Top - r2.Top));
return new Point(x, y);
```

File: MouseUnSnag/ScreenHandling/DisplayList.cs
Comment (Robin-S) [inline MouseUnSnag/ScreenHandling/DisplayList.cs:94]: Running this code, I notice that the cursor is having difficulty moving along the top of a display where there is not another display above it. The debug statements reveal a lot of 'jumps' of the mouse to the `curScreen`:

```1183: StuckDirection/Distance{X=0,Y=-1}/{X=0,Y=-1} cur_mouse:{X=2559,Y=-1}  prev_mouse:{X=2565,Y=-1} ==? cursor:{X=2560,Y=0} (OnMon#2/)  #UnSnags 1179   --STUCK--   
  > JumpScreen
{X=2560,Y=0} -> {X=2560,Y=0}
1184: StuckDirection/Distance{X=0,Y=-1}/{X=0,Y=-1} cur_mouse:{X=2554,Y=-1}  prev_mouse:{X=2559,Y=-1} ==? cursor:{X=2555,Y=0} (OnMon#2/)  #UnSnags 1180   --STUCK--   
  > JumpScreen
{X=2555,Y=0} -> {X=2555,Y=0}
1185: StuckDirection/Distance{X=0,Y=-1}/{X=0,Y=-1} cur_mouse:{X=2552,Y=-1}  prev_mouse:{X=2554,Y=-1} ==? cursor:{X=2553,Y=0} (OnMon#2/)  #UnSnags 1181   --STUCK--   
  > JumpScreen
{X=2553,Y=0} -> {X=2553,Y=0}
```

I think because `curScreen` itself is being considered as a `candidate` for jumping the cursor?

File: MouseUnSnag/GeometryUtil.cs
Comment (Robin-S) [reply MouseUnSnag/GeometryUtil.cs:105]: Consider calculating `r2 - r1` instead of `r1 - r2` for consistency with other direction-methods in this class.

**Comment by Robin-S** [comment]
I think this is a great improvement :)

Please do be aware that this change makes some cursor-jumps look very much like corner-unstick.
Moving the mouse perfectly horizontally onto an existing display would trigger the unstick-functionality.
Movint the mouse diagonally onto a non-existing display would trigger the jump functionality.
image

This could potentially be confusing to users that do not have the 'Unstick from corners' option enabled.

File: MouseUnSnag/GeometryUtil.cs
Comment (GrantMoyer) [reply MouseUnSnag/GeometryUtil.cs:105]: > Consider calculating `r2 - r1` instead of `r1 - r2`

Yes, `r1 - r2` is definitely backward.

File: MouseUnSnag/ScreenHandling/DisplayList.cs
Comment (GrantMoyer) [reply MouseUnSnag/ScreenHandling/DisplayList.cs:94]: Thanks for testing a build of this. Yep, there's a bug in ScreensInDirection.

```c#
return screenDir.X == dir.X && Math.Abs(screenDir.Y - dir.Y) <= 1.0
    || screenDir.Y == dir.Y && Math.Abs(screenDir.X - dir.X) <= 1.0;
```

should be

```c#
return screenDir.X != 0 && screenDir.X == dir.X && Math.Abs(screenDir.Y - dir.Y) <= 1
    || screenDir.Y !- 0 && screenDir.Y == dir.Y && Math.Abs(screenDir.X - dir.X) <= 1;
```

**Comment by GrantMoyer** [comment]
> Please do be aware that this change makes some cursor-jumps look very much like corner-unstick. Moving the mouse perfectly horizontally onto an existing display would trigger the unstick-functionality. Movint the mouse diagonally onto a non-existing display would trigger the jump functionality.

I pushed a change which disables diagonal jumps if unstick is disabled. Since it seems like you regularly use the jump feature, would you mind testing if this matches your intuition?

**Comment by Robin-S** [comment]
> @Robin-S I pushed a change which disables diagonal jumps if unstick is disabled. Since it seems like you regularly use the jump feature, would you mind testing if this matches your intuition?

I have been using this build today with the 4-screen setup seen in #16 and it does indeed match the intuition better :)

**Comment by GrantMoyer** [comment]
Considering the manual testing we've done, I don't think we need to wait for #15.

**Comment by patricknelson** [comment]
Please call me out and LMK if anything needs a rubber stamp 😄 (right now I'm sort of swamped but I know at the moment this repo requires multiple reviewers for approvals)

**Review by Robin-S — APPROVED** [review]

**Comment by GrantMoyer** [comment]
@patricknelson Ping for rubber stamp on this issue, then.

**Review by patricknelson — APPROVED** [review]
LGTM

**Comment by patricknelson** [comment]
Alright! Here goes noth....

BeneficialDecisiveAtlanticblackgoby-max-1mb

**Comment by patricknelson** [comment]
Who's responsible for building the latest release and announcing it on our releases page? Also note that we have some milestones that should still be met (including some minor documentation things which should be done so users actually understand the changes since the last release).

Unfortunately I don't have the bandwidth to take care of that right this particular moment, but in case someone wants to tackle that. See: https://github.com/MouseUnSnag/MouseUnSnag/issues?q=is%3Aopen+is%3Aissue+milestone%3Av2.2

**Pull request merged**
